File: cpp/129.cpp
#include <iostream>

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  void helper(TreeNode* cur, int curSum, int &total) {
    curSum = 10*curSum + cur->val;
    if (cur->left)
      helper(cur->left, curSum, total);
    if (cur->right)
      helper(cur->right, curSum, total);
    if (!cur->left && !cur->right)
      total += curSum;
  }
  int sumNumbers(TreeNode *root) {
    int ret = 0;
    helper(root, 0, ret);
    return ret;
  }
};

File: go/129.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func dfs(cur *TreeNode, curSum int, ret *int) {
  if cur == nil {
    return
  }
	curSum = 10*curSum + cur.Val
	if cur.Left == nil && cur.Right == nil {
	  *ret += curSum
	  return
	}
	dfs(cur.Left, curSum, ret)
	dfs(cur.Right, curSum, ret)
}

func sumNumbers(root *TreeNode) int {
	ret := 0
	dfs(root, 0, &ret)
	return ret
}
